### PR TITLE
dashboard backend

### PR DESCRIPTION
**Pull request by delan**
This patch adds the backend for an intermittents dashboard including:

* sqlite-based storage for test statistics and attempt results
* GET /dashboard/tests, for listing tests with unexpected results
    * ?group to group subtests by path, max unexpected count, max last unexpected
* GET /dashboard/attempts endpoint, for listing unexpected results for a test
    * ?path= to filter by path
    * ?subtest= to filter by subtest (omit param entirely for NULL)
    * ?branch=&build_url=&pull_url= to filter by any of those
    * ?since= to skip rows older than the given time (for live updates)
* POST /dashboard/query endpoint, for filtering intermittents in a single request
* POST /dashboard/attempts endpoint, for storing new attempt results
    * response format is the same as in POST /dashboard/query

## Testing

```
$ printf \%s '[{"title": "x", "number": 1}]' > data/issues.json

$ curl -fsS 'http://127.0.0.1:5000/dashboard/query' -H 'Content-Type:application/json' \
> -d '[{"path":"x","subtest":"b","expected":"FAIL","actual":"PASS"},{"path":"x","subtest":"c","expected":"PASS","actual":"PASS"}]'

{"known": [{"path": "x", "subtest": "b", "issues": [{"title": "x", "number": 1}]}, {"path": "x", "subtest": "c", "issues": [{"title": "x", "number": 1}]}], "unknown": []}

$ curl -fsS 'http://127.0.0.1:5000/dashboard/attempts' -H 'Content-Type:application/json' \
> -d '[{"path":"x","subtest":"b","expected":"FAIL","actual":"PASS"},{"path":"x","subtest":"c","expected":"PASS","actual":"PASS"}]'

curl: (22) The requested URL returned error: 401

$ curl -fsS 'http://127.0.0.1:5000/dashboard/attempts' -H 'Content-Type:application/json' \
> -H 'Authorization: Bearer ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopq' \
> -d '[{"path":"x","subtest":"b","expected":"FAIL","actual":"PASS"},{"path":"x","subtest":"c","expected":"PASS","actual":"PASS"}]'

{"known": [{"path": "x", "subtest": "b", "issues": [{"title": "x", "number": 1}]}, {"path": "x", "subtest": "c", "issues": [{"title": "x", "number": 1}]}], "unknown": []}

$ curl -fsS 'http://127.0.0.1:5000/dashboard/tests' | jq .

[
  {
    "rowid": 1,
    "path": "x",
    "_subtest": "=b",
    "subtest": "b",
    "unexpected_count": 1,
    "last_unexpected": 1675334475,
    "issues": [
      {
        "title": "x",
        "number": 1
      }
    ]
  }
]

$ curl -fsS 'http://127.0.0.1:5000/dashboard/attempts' -H 'Content-Type:application/json' \
> -H 'Authorization: Bearer ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopq' \
> -d '[{"path":"x","subtest":"c","expected":"PASS","actual":"FAIL"},{"path":"z","expected":"PASS","actual":"FAIL"}]'

{"known": [{"path": "x", "subtest": "c", "issues": [{"title": "x", "number": 1}]}], "unknown": [{"path": "z", "subtest": null}]}

$ curl -fsS 'http://127.0.0.1:5000/dashboard/tests' | jq .

[
  {
    "rowid": 2,
    "path": "x",
    "_subtest": "=c",
    "subtest": "c",
    "unexpected_count": 1,
    "last_unexpected": 1675334797,
    "issues": [
      {
        "title": "x",
        "number": 1
      }
    ]
  },
  {
    "rowid": 3,
    "path": "z",
    "_subtest": "",
    "subtest": null,
    "unexpected_count": 1,
    "last_unexpected": 1675334797,
    "issues": []
  },
  {
    "rowid": 1,
    "path": "x",
    "_subtest": "=b",
    "subtest": "b",
    "unexpected_count": 1,
    "last_unexpected": 1675334475,
    "issues": [
      {
        "title": "x",
        "number": 1
      }
    ]
  }
]
```

## Migrating

To prepare the production instance, generate a `dashboard_secret` as shown in the README, and rename `token` in config.json to `github_token`.

### CODE REVIEW

File: schema/dashboard.1.sql [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+CREATE TRIGGER "test.count_unexpected" INSERT ON "attempt" BEGIN
+    INSERT INTO "test" VALUES (
+        new."path", new."subtest"
+        , CASE WHEN new."actual" != new."expected" THEN 1 ELSE 0 END
+        , CASE WHEN new."actual" != new."expected" THEN new."time" ELSE NULL END
+    ) ON CONFLICT DO UPDATE SET
+        "unexpected_count" = "unexpected_count" + 1
+        , "last_unexpected" = new."time"
+        WHERE new."actual" != new."expected";
+END;
```
Comment (delan) [inline schema/dashboard.1.sql]: this was fun to write, but doing logic in triggers is awkward, so i might convert this to python

Comment (delan) [reply schema/dashboard.1.sql]: done in e296a12309f68810cc3227787041ca1bfed536cf

**Review by mrobinson — APPROVED** [review]
This looks great!

**Review by mrobinson — CHANGES_REQUESTED** [review]
Oh, there is one small change that would be useful here. Could the `attempts` endpoint return a list of intermittents that it hasn't seen before from the set provided? This will allow the CI to print whether a flake is new or already known -- making it easier to detect if a change adds flakiness.

**Comment by delan** [comment]
Thanks! POST response done in 7dc4c7a86beb7d09cafa60f25f26f85264fbeda4.

**Review by mrobinson — APPROVED** [review]
Great! Thank you.

**Comment by delan** [comment]
Migration steps for #4 and #5 done in production. The service will fail to restart unless updated to include this patch, due to the lack of `token` in config.json.

File: schema/dashboard.1.sql [hunk 2]
```diff
@@ -0,0 +1,32 @@
+CREATE TABLE "meta" (
+    "version" INTEGER NOT NULL
+);
+
+-- dashboard.2.sql should UPDATE "meta" SET "version" = 2; and so on
+INSERT INTO "meta" VALUES (1);
+
+CREATE TABLE "test" (
+    "path" TEXT NOT NULL                        -- test name
+    , "subtest" TEXT
```
Comment (delan) [inline schema/dashboard.1.sql]: tests are identified by (path,subtest), but some tests are not subtests. the obvious representation for such tests would be (path,NULL), but sqlite makes the (legal) decision to allow (path,NULL) to effectively bypass

1. PRIMARY KEY constraints ([“Each row in a table with a primary key must have a unique combination of values in its primary key columns. For the purposes of determining the uniqueness of primary key values, NULL values are considered distinct from all other values, including other NULLs.”](https://sqlite.org/lang_createtable.html#the_primary_key))
2. UNIQUE constraints ([“For each UNIQUE constraint on the table, each row must contain a unique combination of values in the columns identified by the UNIQUE constraint. For the purposes of UNIQUE constraints, NULL values are considered distinct from all other values, including other NULLs.”](https://sqlite.org/lang_createtable.html#unique_constraints))
3. FOREIGN KEY constraints ([“The foreign key constraint is satisfied if for each row in the child table either one or more of the child key columns are NULL, or there exists a row in the parent table for which each parent key column contains a value equal to the value in its associated child key column.”](https://sqlite.org/foreignkeys.html#fk_basics))

Comment (delan) [reply schema/dashboard.1.sql]: i managed to solve this by converting (path,NULL) to (path,'') and (path,subtest) to (path,'='||subtest) between the application and the database (057f6b095ed5f0ab2502e30d1707781173b76b39), but this makes every insert and select a bug waiting to happen.

tomorrow i’ll see if i can automate this conversion with a [generated column](https://sqlite.org/gencol.html) that gets used in constraints.

Comment (delan) [reply schema/dashboard.1.sql]: done in 929854f14d2081731d9a99d7cd70065644ed763a

**Review by mrobinson — APPROVED** [review]
This looks good to me!

**Pull request merged**